codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/AsyncClientClass.java

```diff
@@ -374,24 +376,29 @@ protected MethodSpec.Builder operationBody(MethodSpec.Builder builder, Operation
                 }
             }
 
-            builder.addStatement("$T cachedEndpoint = null", URI.class);
+            builder.addStatement("$T<$T> endpointFuture = $T.completedFuture(null)",
+                                 CompletableFuture.class, URI.class, CompletableFuture.class);
             builder.beginControlFlow("if (endpointDiscoveryEnabled)");
 
-            builder.addCode("$T key = $N.overrideConfiguration()", String.class, opModel.getInput().getVariableName())
+            ParameterizedTypeName identityFutureTypeName = ParameterizedTypeName.get(ClassName.get(CompletableFuture.class),
+                                                             WildcardTypeName.subtypeOf(AwsCredentialsIdentity.class));
+            builder.addCode("$T identityFuture = $N.overrideConfiguration()", identityFutureTypeName,
+                            opModel.getInput().getVariableName())
                    .addCode("    .flatMap($T::credentialsIdentityProvider)", AwsRequestOverrideConfiguration.class)
                    .addCode("    .orElseGet(() -> clientConfiguration.option($T.CREDENTIALS_IDENTITY_PROVIDER))",
                             AwsClientOption.class)
-                   // TODO: avoid join inside async
-                   .addCode("    .resolveIdentity().join().accessKeyId();");
```

Comment:
This PR only changes the handling in AsyncClientClass.java. A join will remain in SyncClientClass.java, though it will change slightly for exception handling - that'll be in a separate PR.